### ISSUE TITLE
ChatBox Issues and Feedback

### ISSUE THREAD

**Issue by MarcSkovMadsen** [holoviz/panel#5135]
panel==1.1.0

I'm trying out the new `ChatBox` widget which is great we now have. Here is my feedback and some issues.

I WILL UPDATE THIS LIST AS I EXPLORE THE `ChatBox` in a real example.

### Does not work well with `sizing_mode="stretch_width"`

I always set `sizing_mode="stretch_width"` in `pn.extension` because I want and need responsive behaviour. 

```python
import panel as pn

pn.extension(sizing_mode="stretch_width", template="fast")

chat_box = pn.widgets.ChatBox(value=[{"bot": "Welcome"}], primary_name="user")

pn.Column(chat_box).servable()
```

I would expect the chat message to be left or right aligned as in the reference guide. Not stretching to full width, which makes it harder to understand the flow.

![image](https://github.com/holoviz/panel/assets/42288570/532f8255-d727-432c-bbfa-9aa1e355d8ba)

### Default colors do not work well with Fast Template

I almost always use the `Fast` templates as their styling is the most mature. But the message colors does not work well with this template as you can see in the example above. The `Chatbox` blue does not work well with the Panel Blue and the `gray` color of the bot is the same as the card background color. This again makes it hard to follow the flow.

### Overflow when setting fixed height with Fast Template

I think its a very natural thing to do to set a fixed height such that there is is space in your application.

```python
import panel as pn

pn.extension(sizing_mode="stretch_width", template="fast")

chat_box = pn.widgets.ChatBox(value=[{"System": "You are now a welcoming assistant"}, {"Assistant": "Welcome"}, {"User": "Thanks"}]*30, primary_name="User", message_colors={
    "User": "#f0cce2", "Assistant": "#cce2f0", "System": "lightgray", 
}, ascending=True, height=600)

pn.Column(chat_box).servable()
```

![image](https://github.com/holoviz/panel/assets/42288570/e4313b64-6a81-4a6e-908e-49335e7a9541)

### name in a confusing position when `ascending=True`

If I set `ascending=True` to get the chat to work like chatGPT and other chats the name of the user is positioned closer to the previous message. I have a feeling that it would also be better to position it below the message it belongs to.

![image](https://github.com/holoviz/panel/assets/42288570/7b5b2219-34cf-4cfe-982a-8d293ef1bd8c)

```python
import panel as pn

pn.extension(sizing_mode="stretch_width", template="fast")

OPEN_AI_CHATBOX_OPTIONS = dict(
    primary_name="User",
    ascending=True,
)

FAST_AI_CHATBOX_OPTIONS = {
    "message_colors": {
        "User": "#f0cce2",
        "Assistant": "#cce2f0",
        "System": "lightgray",
    },   
}

chat_box = pn.widgets.ChatBox(
    value=[
        {"System": "You are now a welcoming assistant"},
        {"Assistant": "Welcome"},
        {"User": "Thanks"},
    ],
    **OPEN_AI_CHATBOX_OPTIONS,
    **FAST_AI_CHATBOX_OPTIONS,
)

pn.Column(chat_box, height=800).servable()
```

It becomes even worse if I add the icons

![image](https://github.com/holoviz/panel/assets/42288570/b4618547-be42-4c80-ade9-8b89173446e0)

```python
import panel as pn

pn.extension(sizing_mode="stretch_width", template="fast")

CHATBOX_OPTIONS = dict(
    primary_name="User",
    ascending=True,
    message_icons={
        "User": "https://user-images.githubusercontent.com/42288570/246667322-33a2a320-9ea3-4e79-8fb8-fcb5b6eac9c0.png",
        "System": "https://user-images.githubusercontent.com/42288570/246667324-5cf26789-765f-4f76-a8bf-49309d2ae84f.png",
        "Assistant": "https://user-images.githubusercontent.com/42288570/246667325-ad4e3434-d173-4463-bb98-5c5d4a892b25.png",
    }
)

FAST_CHATBOX_OPTIONS = {
    "message_colors": {
        "User": "#f0cce2",
        "Assistant": "#cce2f0",
        "System": "lightgray",
    },

}

chat_box = pn.widgets.ChatBox(
    value=[
        {"System": "You are now a welcoming assistant"},
        {"Assistant": "Welcome"},
        {"User": "Thanks"},
    ],
    **CHATBOX_OPTIONS,
    **FAST_CHATBOX_OPTIONS,
)

pn.Column(chat_box, height=800).servable()

```

### Scroll to top/ bottom not working

Its not working for me. If you look at the implementation you will se that if it works at all it will only work when the only `Column`s you include in your app is from one ChatBox.

![image](https://github.com/holoviz/panel/assets/42288570/9aee53db-223c-4cec-9062-54dee6c11b56)

I also tried to figure out how I could use the `chat_log` in `args` or something similar. But it seems I cannot get from that to the HTML element and the scroll that one.

I think it would be quite useful in general to add `scroll_to_top` and `scroll_to_bottom` methods to a Column with a scroll bar.

I think the *scroll* functionality would be nice to have on a Column in general. So I've added a Feature Request https://github.com/holoviz/panel/issues/5137.

### Dictionary is not the right container for messages.

The `value` of the `ChatBox` is a list of dictionaries.

```python
 value=[
        {"System": "You are now a welcoming assistant"},
        {"Assistant": "Welcome"},
        {"User": "Thanks"},
 ```

I believe this is not the right format for *messages. Its really hard to work with and extract the *user* `name` or *message* `text`.

I believe the format should be one of these two

```python
 value=[
        {"user": "System", "message": "You are now a welcoming assistant"},
        ...
]
```

or

```python
 value=[
        ("System", "You are now a welcoming assistant"),
        ...
]
```

## `ChatBox` is too slow to work with after +20 elements have been added.

This is really a problem for the `Column` that I've reported here https://github.com/holoviz/panel/issues/5136

### Change `primary_name` to `primary_user`

As a new user of the `ChatBox` the `primary_name` parameter is not self explanatory - *name* can refer to anything. I then have to look it up in the documentation. Something like `primary_user` would have been self explanatory for me.

Please consider changing the parameter name.

### Provide example using message_icons

Its a very natural thing to want to set the `message_icons`, but the Reference Guide does not contain an example. So using `message_icons` takes time. I have to

- Understand how it works.
- Find some relevant icons to use and place them in my project. Could take 15 mins.

Please include an example using `message_icons` and please provide easy to use icons for the `System`, `Assistant` and `User` profiles that OpenAI are moving to as their standard. Icons like the below

<img src="https://github.com/holoviz/panel/assets/42288570/33a2a320-9ea3-4e79-8fb8-fcb5b6eac9c0" width="25px"></img>

<img src="https://github.com/holoviz/panel/assets/42288570/5cf26789-765f-4f76-a8bf-49309d2ae84f" width="25px"></img>

<img src="https://github.com/holoviz/panel/assets/42288570/ad4e3434-d173-4463-bb98-5c5d4a892b25" width="25px"></img>

<img src="https://github.com/holoviz/panel/assets/42288570/d3a26763-f7f5-4e8d-8933-cb69670f90a8" width="25px"></img>

<img src="https://github.com/holoviz/panel/assets/42288570/70c35e2e-c58e-4cb7-9595-4665bf202952" width="25px"></img>

### Simplify User Experience to ChatGPT simplicity

Please consider if the current layout is not to difficult to get right and too hard to use for users.

Why not just do what ChatGPT does. That is what every user would expect.

- icons on the left
- messages on the center
- actions on the right
- Simple divider between messages

![image](https://github.com/holoviz/panel/assets/42288570/b3d80a5b-9892-4fc5-8a12-0fd6c62d6344)

I think having messages alternating between left and right makes it hard to read as a document.

![image](https://github.com/holoviz/panel/assets/42288570/ff0182b0-4d9e-4a42-a793-a7e6509904b0)

At least consider. If the issues can be fixed this feature request might not be nescessary.

## Enable me to not show the name

When I have icons as in the example above I would not want to show the `User`, `Assistant` and `System` name. Please enable me to hide it. Or just do it if there is an icon.

## Enable me to change the default value of `show_name` and other parameters

With the current implementation this is not possible because the value is set to `True` in the constructor.

Then I cannot use `pn.widgets.chatbox.ChatRow.param.show_name.default=False`.

This is going against how Param and Panel objects are supposed to work.

![image](https://github.com/holoviz/panel/assets/42288570/cd0333eb-f5ce-4283-9417-d00eec2061e7)

**Comment by MarcSkovMadsen** [comment]
## Alternative Version

I need a working chat interface. I need it to work for me and my use case. So therefore I have to make own version. I will share it here so that we can align or maybe merge it into the Panel `ChatBox` when its matured.

I will update this example when it makes sense.

Status. PROOF OF CONCEPT

```python
import panel as pn

pn.extension(sizing_mode="stretch_width", template="fast")

CHAT_ICONS = {
    "User": "https://user-images.githubusercontent.com/42288570/246667322-33a2a320-9ea3-4e79-8fb8-fcb5b6eac9c0.png",
    "System": "https://user-images.githubusercontent.com/42288570/246671694-70c35e2e-c58e-4cb7-9595-4665bf202952.png",
    "Assistant": "https://user-images.githubusercontent.com/42288570/246671017-d3a26763-f7f5-4e8d-8933-cb69670f90a8.svg",
}
CHAT_COLORS = {
    "User": "#e5f0f7",
    "Assistant": "var(--neutral-stroke-divider-rest)",
    "System": "var(--neutral-stroke-focus",
}

EXAMPLE_MESSAGES = [{"System": "You are now a welcoming assistant"}] + [
    {"Assistant": "Welcome"},
    {"User": "Thanks"},
] * 15

INITIAL_SCRIPT = """
<img src onerror="(()=>{
this.onerror = null;// prevent endless loop if function generates an error
var column = null
if (window.el){
    column = window.el;
} else {
    const element = this.getRootNode().host;
    column = element.getRootNode().host;
    window.el=column;
}
column.scrollTop = column.scrollHeight;
})()">
"""

index={"value": 0}

max_rows = 10
rows = []

script = pn.pane.HTML(INITIAL_SCRIPT, sizing_mode="fixed", width=0, height=0, margin=0)
def trigger_scroll():
   
   script.object=f"""<div style="display:none">{index["value"]}</div>""" + INITIAL_SCRIPT

def create_chat_row(*objects, user: str = "User"):
    value = index["value"]
    index["value"]+=1
    return pn.Row(
        pn.Column(
            pn.pane.Image(
                object=CHAT_ICONS[user],
                height=25,
                width=25,
                sizing_mode="fixed",
                align=("center", "center"),
                margin=(30, 10, 0, 5),
                embed=False,
            ),
            pn.pane.Str(user, margin=(0, 10, 0, 5), stylesheets=["pre {text-align: center;}"]),
            sizing_mode="fixed", width=100
        ),
        pn.Column(
            *objects,
            sizing_mode="stretch_both",
            styles={"background": CHAT_COLORS[user], "border-radius": "5px"},
            margin=(10, 5),
        ),
        sizing_mode="stretch_width",
    )


prompt = pn.widgets.TextInput(
    sizing_mode="stretch_width", placeholder="Submit a task", margin=(10, 23, 10, 105), name="Press ENTER to submit the task"
)



chat_messages = pn.Column(
    script, height=600, styles={"overflow-y": "scroll"}, css_classes=["chat-messages"]
)


show_all_messages = pn.widgets.Button(name="Show More", align=("center", "center"), sizing_mode="fixed", width=200, margin=(10, 23, 10, 105), button_type="primary", button_style="outline")

@pn.depends(show_all_messages, watch=True)
def show_messages(show):
    chat_messages[:]=[script, *rows]


def add_message(*objects, user: str = "User") -> pn.Row:
    row = create_chat_row(*objects, user=key)
    rows.append(row)
    
    if len(rows)<=max_rows:
        chat_messages[:]=[script, *rows]
    else:
        chat_messages[:]=[script, show_all_messages, *rows[-10:]]
    trigger_scroll()
    return row


for message in EXAMPLE_MESSAGES:
    for key, value in message.items():
        add_message(value, user=key)


@pn.depends(prompt.param.value, watch=True)
def send_message(value):
    if value!="":
        print("send_message")
        add_message(value, user="User")
        prompt.value = ""

chat_box = pn.Column(
    chat_messages,
    prompt,
).servable()
```

**Comment by ahuang11** [comment]
Thanks for bringing these issues to light.

I addressed the layout issues here:
https://github.com/holoviz/panel/pull/5139

It'll take more time for me to address the other issues

**Comment by ahuang11** [comment]
[ ] Does not work well with sizing_mode="stretch_width"
[ ] Default colors do not work well with Fast Template
[ ] Overflow when setting fixed height with Fast Template
[ ] name in a confusing position when ascending=True
[ ] Scroll to top/ bottom not working
[ ] Dictionary is not the right container for messages.
[ ] ChatBox is too slow to work with after +20 elements have been added.
[ ] Change primary_name to primary_user
[ ] Provide example using message_icons
[ ] Simplify User Experience to ChatGPT simplicity
[ ] Enable me to not show the name
[ ] Enable me to change the default value of show_name and other parameters

**Comment by MarcSkovMadsen** [comment]
## Value order depends on `ascending` parameter.

I can see that the order or the message `value`s depends on the value of the `ascending` parameter. I believe this is a design mistake as it makes it harder to use items from `value` as you will have to take `ascending` into account when you want to take the first or last message added.

**Comment by ahuang11** [comment]
I do not believe that is true. Ascending just sets flex-direction column reverse; the items are still appended at the bottom.

**Comment by MarcSkovMadsen** [comment]
I will close this one as the `ChatBox` is about to be deprecated with Panel 1.3.0.
